### ISSUE TITLE
Make `ConstraintSynthesizer::generate_constraints` not consume `self`

### ISSUE THREAD

**Issue by Pratyush** [arkworks-rs/snark#6]

**Comment by weikengchen** [comment]
Does it mean that we just change `self` to `&self`?

My first attempt shows that this seems good: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0dcc9126df1c78347ad34fc5238617a4

**Comment by Pratyush** [comment]
yeah, it's mostly just an issue with updating everything downstream =)
